keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.onSucceededPods.indexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `indexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	Indexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
+	// - Terminate: indicates that not finished pods would be terminated.
+	WhenTargetAchieved WhenTargetAchievedSuccessPolicy
```

Comment:
I think it is more natural to put `WhenTargetAchieved` directly under `SuccessPolicy`. The field `OnSucceededPods` would still describe the target.

Reply:
Given that we add a new `successPolicy` so that we can configure the `whenTargetAchived` in each `successPolicy`, I think that this API form (put the `whenTargetAchived` under the `onSuceededPods`) would be better.

WDYT? @mimowo 

Anyway, we need to review new APIs by API reviewers.

Reply:
Not sure I understand `Given that we add a new successPolicy so that we can configure the whenTargetAchived in each successPolicy`, sounds very abstract. Do you have a particular extension of the API that would be harder if the `whenTargetAchieved` what just under `successPolicy`?

In my understanding, we have just one `successPolicy` in the API. If we intend to support having multiple, maybe we should think of a list of rules like in pod failure policy, but I'm not sure we want to go this path.

Also, I find this API a little bit confusing, because it is not clear what the target is, this is why I propose the change as here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1453297580.

Reply:
> Do you have a particular extension of the API that would be harder if the whenTargetAchieved what just under successPolicy?

@mimowo We can consider the following situations in which we add new success policies in the future:

```yaml
...
spec:
  successPolicy:
    onPolicyA:
      paramA0: ""
    onPolicyB:
      paramB0: ""
    onSucceededPods:
      count: 5
      whenTargetAchieved: terminate
...
```

If we directly put `whenTargetAchieved` under the `.spec.successPolicy` field, the `whenTargetAchived` affects other successPolicies.

> In my understanding, we have just one successPolicy in the API. If we intend to support having multiple, maybe we should think of a list of rules like in the pod failure policy, but I'm not sure we want to go this path.

Defining the successPolicy as a form similar to the pod failure policy sounds attractive.
However, I'm not sure how extensible we should define APIs for the future.
 Let me know what you think. @deads2k

> Also, I find this API a little bit confusing, because it is not clear what the target is, this is why I propose the change as here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1453297580.

It sounds reasonable to me.
Maybe we should have the `action` field instead of the `whenTargetAchived` like this:

```
  successPolicy:
    onSucceededPods:
      count: 5
      action: terminateRunning | none
```

The `action` means the following semantics:

- `terminateRunning`: All pods, including running one, are terminated once the job meets this successPolicy. 
- `none`: Even if the job meets this successPolicy, running pods aren't terminated.

Reply:
Related thread is here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1460681741

Reply:
We want to continue discussing API design here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1458041856